cmd/legacy/heartbeat/params.go

```diff
@@ -344,128 +334,97 @@ func parseExtraHeartbeat(data string) ([]heartbeat.Heartbeat, error) {
-		Timestamp         *string            `json:"timestamp"`
-	}
-
-	err := json.Unmarshal([]byte(data), &incoming)
-	if err != nil {
-		return nil, fmt.Errorf("failed to json decode from data %q: %s", string(data), err)
-	}
-
-	var heartbeats []heartbeat.Heartbeat
-
-	for _, h := range incoming {
 		var cursorPosition *int
 
-		if h.CursorPosition != nil {
-			parsed, err := strconv.Atoi(*h.CursorPosition)
+		switch cursorPositionVal := h.CursorPosition.(type) {
+		case float64:
+			cursorPosition = heartbeat.Int(int(cursorPositionVal))
+		case string:
+			val, err := strconv.Atoi(cursorPositionVal)
 			if err != nil {
 				return nil, fmt.Errorf("failed to convert cursor position to int: %s", err)
 			}
 
-			cursorPosition = &parsed
-		}
-
-		h.Entity, err = homedir.Expand(h.Entity)
-		if err != nil {
-			return nil, fmt.Errorf("failed expanding entity: %s", err)
+			cursorPosition = heartbeat.Int(val)
 		}
 
-		var entityType heartbeat.EntityType
+		var isWrite *bool
 
-		// Both type or entity_type are acceptable here. Type takes precedence.
-		entityTypeStr := firstNonEmptyString(h.Type, h.EntityType)
-		if entityTypeStr != "" {
-			entityType, err = heartbeat.ParseEntityType(entityTypeStr)
+		switch isWriteVal := h.IsWrite.(type) {
+		case bool:
+			isWrite = heartbeat.Bool(isWriteVal)
+		case string:
+			val, err := strconv.ParseBool(isWriteVal)
 			if err != nil {
-				return nil, err
+				return nil, fmt.Errorf("failed to convert is write to bool: %s", err)
 			}
+
+			isWrite = heartbeat.Bool(val)
 		}
 
 		var lineNumber *int
 
-		if h.LineNumber != nil {
-			parsed, err := strconv.Atoi(*h.LineNumber)
+		switch lineNumberVal := h.LineNumber.(type) {
+		case float64:
+			lineNumber = heartbeat.Int(int(lineNumberVal))
+		case string:
+			val, err := strconv.Atoi(lineNumberVal)
 			if err != nil {
 				return nil, fmt.Errorf("failed to convert line number to int: %s", err)
 			}
 
-			lineNumber = &parsed
+			lineNumber = heartbeat.Int(val)
 		}
 
 		var lines *int
 
-		if h.Lines != nil {
-			parsed, err := strconv.Atoi(*h.Lines)
+		switch linesVal := h.Lines.(type) {
+		case float64:
+			lines = heartbeat.Int(int(linesVal))
+		case string:
+			val, err := strconv.Atoi(linesVal)
 			if err != nil {
 				return nil, fmt.Errorf("failed to convert lines to int: %s", err)
 			}
 
-			lines = &parsed
+			lines = heartbeat.Int(val)
 		}
 
-		var timestamp float64
+		var time float64
 
-		switch {
-		case h.Time != nil:
-			parsed, err := strconv.ParseFloat(*h.Time, 64)
+		switch timeVal := h.Time.(type) {
+		case float64:
+			time = timeVal
+		case string:
+			val, err := strconv.ParseFloat(timeVal, 64)
 			if err != nil {
 				return nil, fmt.Errorf("failed to convert time to float64: %s", err)
 			}
 
-			timestamp = parsed
-		case h.Timestamp != nil:
-			parsed, err := strconv.ParseFloat(*h.Timestamp, 64)
+			time = val
+		}
+
+		var timestamp float64
+
+		switch timestampVal := h.Timestamp.(type) {
+		case float64:
```

Comment:
For `timestamp` property, support for case `int` type would make it more robust as well.